### ISSUE TITLE
[server] Type name X redefined from module X

### ISSUE THREAD

**Issue by Antriel** [HaxeFoundation/haxe#8368]
There's been a lot of talk about this type of error. I'm not sure if it's one issue or multiple, but I managed to make a minimal example for my current one. Interestingly it's very similar to https://github.com/HaxeFoundation/haxe/issues/8043#issuecomment-491915152 from my previous issue.
hxml
```hxml
-lib hxnodejs #this isn't really used, but without it, it works
-cp src
-main Main
-js build.js
```
Main.hx
```hx
class Main {
    public static function main():Void {
        MyMacro.foo();
    }
}
```
MyMacro.hx
```hx
class MyMacro {
    #if macro
    static var issue = haxe.macro.ComplexTypeTools.toType(macro :js.lib.Uint8Array);
    // ^ this causes the issue
    #end
    public static macro function foo() return macro { };
}
```
Compiling this normally works. First compile via `haxe --connect 6000 build.hxml` works, second one will result in `C:\HaxeToolkit\std/js/lib/Uint8Array.hx:28: characters 36-51 : Type name js.lib.Uint8Array is redefined from module js.lib.Uint8Array`.

It's the combination of the `ComplexTypeTools.toType` and the hxnodejs library (tested with haxelib latest, 6.9.1).
Haxe version https://github.com/HaxeFoundation/haxe/commit/877addced238fbf2cfdd3010e19f3b906d777aa8.

Possibly related to https://github.com/HaxeFoundation/haxe/issues/6993, https://github.com/HaxeFoundation/haxe/issues/4234, https://github.com/HaxeFoundation/haxe/issues/5725. Maybe also related to https://github.com/HaxeFoundation/haxe/issues/7279, I'm not sure as that seems to complain more about the error reporting than the error being false.

**Comment by kLabz** [comment]
> `-lib hxnodejs #this isn't really used, but without it, it works`

You can replace it with `--macro define('nodejs')`

**Comment by Antriel** [comment]
> --macro define('nodejs')

Indeed that works (i.e. causes the issue) too.

**Comment by kLabz** [comment]
A variant of the exact same issue:

build.hxml
```hxml
# Just need any --macro call
--macro define('whatever')
-main Main
-D no-compilation
```

Main.hx
```hx
class Main {
	public static function main():Void {
		MyMacro.foo();
	}
}
```

MyMacro.hx
```hx
class MyMacro {
	#if macro
	static var issue = haxe.macro.ComplexTypeTools.toType(macro :Type1);
	// ^ this causes the issue
	#end

	public static macro function foo() return macro { };
}
```

Type1.hx
```hx
class Type1 {
	public function new() {
		trace(Type2);
	}
}
```

Type2.hx
```hx
class Type2 {
	public function new() {
		trace(Type1);
	}
}
```

**Comment by kLabz** [comment]
I have a fix.. now to see if it doesn't break everything else :D

**Comment by Antriel** [comment]
https://github.com/HaxeFoundation/haxe/pull/8371 seems to have fixed this issue in my project.

**Comment by RealyUniqueName** [comment]
Some logs from my investigation (based on [this sample](https://github.com/HaxeFoundation/haxe/issues/8368#issuecomment-498606497)):

Issue is reproducible if `ComplexTypeTools.toType(macro:Type1)` is used as an init expr for a static var (1).
If that expr is invoked in a body of a macro function, then everything works fine (2).

In (1)  the module "Type1" is loaded with `macro` and `eval` defined and then loaded again as a dependency of "Type2" without `eval` and `macro` defined, which leads to two different cache signatures for module "Type1". That makes "Type1" to be loaded twice.

While in (2) the module "Type1" is always loaded without `macro` and `eval` defined.

So, I assume that for init expressions of static fields  in macros the context is not properly set up.

Also, In the sample "Type1" is always loaded with `ctx.m.curmod` being `null_module`.
But if I move `ComplexTypeTools.toType(macro:Type1)` from static field init expr to the body of a macro function, then "Type1" is loaded with `ctx.m.curmod` being `Main`, and everything works fine.
That could happen because function execution is delayed more than execution of static init expressions.

**Comment by lublak** [comment]
The same problem apears if i use `Context.defineType`.
With: https://gist.github.com/nadako/2ad4246f257e627a5833
I also use "@:persistent" at this var.
But it shows not everytime. I don't know. How to reproduce it.
With defineModule the "error" shows more often.
For me i use `Context.getType(path) == null` to check if the type is defined before. (But i think it is not the best solution)

**Comment by kLabz** [comment]
Yeah in my project too it's the combo `Context.defineType` + `Context.getType` that triggers this.

**Comment by RealyUniqueName** [comment]
#8571 fixes samples provided in this issue. 
But a more generic fix could only be done after #7871
I suggest to close this issue regardless the destiny of #8571 and focus on #7871

**Comment by RealyUniqueName** [comment]
I'm reverting the fix, because it makes us to add more and more hacks. We will look for a better solution.

**Comment by Simn** [comment]
Here's my preliminary conclusion on what goes wrong in general:

## Compilation 1

### signature A
* An init macro causes a type to be loaded in typing context, let's call that type `Ref[A]`.
* An init macro calls `Compiler.define`, which causes a signature change (hxnodejs does this).

### signature B
* Typing continues, a type `Type[B]` is loaded and references `Ref[A]`, which causes `Ref[A]` to become one of its dependencies.
* Typing finishes, both `Type[B]` and `Ref[A]` are stored in the cache.

## Compilation 2

### signature A
* For whatever reason, `Ref` is _not_ loaded into the typing context immediately, so there is no `ref[A]` there.
* The signature changes again, from an init macro or through `-D`.

### signature B
* Now `Ref` is loaded and becomes `Ref[B]`.
* `Type[B]` is retrieved from the cache.
* Its dependencies are also retrieved from the cache. But `Type[B]` has `Ref[A]` as its dependency.
* `Ref[B]` vs. `Ref[A]`: "Type name Ref is redefined from module Ref"

**Comment by stnguyen** [comment]
Is this fixed in Haxe 4.0.0?

**Comment by RealyUniqueName** [comment]
No

**Comment by Gama11** [comment]
Apparently @nadako's signal macro reproduces this as well, perhaps useful in future investigations of this: https://gist.github.com/nadako/b086569b9fffb759a1b5#gistcomment-3164185

**Comment by RealyUniqueName** [comment]
I still don't know how to fix this properly.
I don't think we can do anything here. Instead we should focus on https://github.com/HaxeFoundation/haxe/issues/7871 and then this issue will become irrelevant.

**Comment by joelmo** [comment]
It seems to be possible to trigger this bug or one thats related. The error message is abit different:
`[Error - 16:03:31] Request textDocument/hover failed.
  Message: Type name haxe.display.JsonTodo is redefined from module haxe.display.JsonModuleTypes
  Code: -32603`

Steps to reproduce:
1. Create a autoBuild for a interface
2. Define new class by reification and also implement the interface (the one that calls the macro):
```
var klass = Context.getLocalClass().get();
// Guard against redefining
if (Context.defined("display") || !klass.isInterface || klass.name == "Client")
	return null;
var client = macro class Client {
        public function new() {}
};
switch client.kind {
	case TDClass(superClass, interfaces, isInterface, isFinal, isAbstract):
		interfaces.push({pack: klass.pack, name: klass.name});
	default:
}
client.pack = ["nocollisionspls"];
Context.defineType(client);
```
4. Try to use autocomplete in vscode. Compiling this works just fine. But if you type x in another line (loads completion suggestions), then remove x and recompile this error appears: `JsonTodo is redefined`. Another issue is that its fields doesnt show up in the suggestions.
```
var x = new nocollisionpls.Client();
x.
```

Offtopic: it would be nice to also have "interpolation" for interfaces in class reification like this `macro class Client implements $klassInterface {}`.

**Comment by kLabz** [comment]
> `// Guard against redefining`

But you're not guarding against redefining, in case of compilation server (which is used for completion but can be used for compilation too) this code can run several times.

**Comment by joelmo** [comment]
Since I'm not in control of the compliation server. Is that a valid scenario you would need to guard against and is it possible in a feasible way? I added a static boolean outside the build function in a attempt to make the build macro only run once, this doesnt work. In this example there should only be one interface that extend the @:autoBuild one.

**Comment by back2dos** [comment]
If you want static bools to persist between compilations, you will need to annotate them with `@:persistent` (although this is not 100% guaranteed to work). As for guarding against redeclaration, you check against the existence of a type with `var exists = try { haxe.macro.Context.getType(name); true; } cache (e:Dynamic) false;`. That said, if `JsonTodo` is the offending type here, there's probably something else going wrong.

**Comment by joelmo** [comment]
After adding `@:persistent` boolean run once check. And a try/catch guard. And one additional "`@:persistent` types already built array check". Problem still occurs.

**Comment by shealyrd** [comment]
In my case, this was being caused by the use of -macro include() in the .hxml. Removing those calls fixed the issue. 

This error was being quietly spit by the completion server and wasn't showing up in compilation. While the completion server was outputting this error, it shut down all completion in the IDE, effectively making the entire project unworkable. 

I know that the vscode extension has its own repo to post in, but while I'm here: I think some priority should be given to stabilizing the IDE, as in my experience it randomly stops functioning for weird, untraceable reasons, which causes big timesinks. Errors in the project will cause the intellisense to stop working, but that's when you would need intellisense the most.

**Comment by Simn** [comment]
@shealyrd What Haxe version is that in your case?

**Comment by shealyrd** [comment]
> @shealyrd What Haxe version is that in your case?

4.2.5, I'll try updating to 4.3 and see if it helps

**Comment by kLabz** [comment]
Yeah this exact issue should be solved with 4.3.0. There are other ways to have the issue, though.

**Comment by Simn** [comment]
I think the original issue has been addressed, but the history here is a mess so I can't really tell. I'll close regardless because I don't want everyone who runs into this particular error to come here.
